pkg/controller/seed-controller-manager/mla/ratelimit_cortex_controller.go

```diff
@@ -0,0 +1,248 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mla
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"go.uber.org/zap"
+	"gopkg.in/yaml.v2"
+
+	kubermaticv1 "k8c.io/kubermatic/v2/pkg/crd/kubermatic/v1"
+	"k8c.io/kubermatic/v2/pkg/kubernetes"
+	"k8c.io/kubermatic/v2/pkg/version/kubermatic"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	runtimeConfigMap      = "cortex-runtime-config"
+	runtimeConfigFileName = "runtime-config.yaml"
+)
+
+type tenantOverride struct {
+	IngestionRate      *int32 `yaml:"ingestion_rate,omitempty"`
+	MaxSeriesPerMetric *int32 `yaml:"max_series_per_metric,omitempty"`
+	MaxSeriesPerQuery  *int32 `yaml:"max_series_per_query,omitempty"`
+	MaxSamplesPerQuery *int32 `yaml:"max_samples_per_query,omitempty"`
+	IngestionBurstSize *int32 `yaml:"ingestion_burst_size,omitempty"`
+	MaxSeriesTotal     *int32 `yaml:"max_series_per_user,omitempty"`
+}
+
+type overrides struct {
+	Overrides map[string]tenantOverride `yaml:"overrides"`
+}
+
+// ratelimitCortexReconciler stores necessary components that are required to manage MLA(Monitoring, Logging, and Alerting) setup.
+type ratelimitCortexReconciler struct {
+	ctrlruntimeclient.Client
+
+	log                       *zap.SugaredLogger
+	workerName                string
+	recorder                  record.EventRecorder
+	versions                  kubermatic.Versions
+	ratelimitCortexController *ratelimitCortexController
+}
+
+// Add creates a new MLA controller that is responsible for
+// managing Monitoring, Logging and Alerting for user clusters.
+func newRatelimitCortexReconciler(
+	mgr manager.Manager,
+	log *zap.SugaredLogger,
+	numWorkers int,
+	workerName string,
+	versions kubermatic.Versions,
+	ratelimitCortexController *ratelimitCortexController,
+) error {
+	client := mgr.GetClient()
+
+	reconciler := &ratelimitCortexReconciler{
+		Client: client,
+
+		log:                       log,
+		workerName:                workerName,
+		recorder:                  mgr.GetEventRecorderFor(ControllerName),
+		versions:                  versions,
+		ratelimitCortexController: ratelimitCortexController,
+	}
+
+	ctrlOptions := controller.Options{
+		Reconciler:              reconciler,
+		MaxConcurrentReconciles: numWorkers,
+	}
+	c, err := controller.New(ControllerName, mgr, ctrlOptions)
+	if err != nil {
+		return err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &kubermaticv1.MLAAdminSetting{}}, &handler.EnqueueRequestForObject{}); err != nil {
```

Comment:
Can we use a predicate to make sure we only enqueue the MLAAdminSetting object with the predefined name: https://github.com/kubermatic/kubermatic/blob/master/pkg/controller/seed-controller-manager/mla/datasource_grafana_controller.go#L101?

Reply:
done